reference/src/representation/enums.md

```diff
@@ -0,0 +1,376 @@
+   A: u32,
+   B: MyEnumPayloadB,
+   C: MyEnumPayloadC,
+   D: (),
+}
+
+#[repr(C)]
+struct MyEnumPayloadB(f32, u64);
+
+#[repr(C)]
+struct MyEnumPayloadC { x: u32, y: u8 }
+}
+```
+
+This enum can also be represented in C++ as follows:
+
+```c++
+#include <stdint.h>
+
+enum class MyEnumTag: CppEquivalentOfInt { A, B, C, D };
+struct MyEnumPayloadB { float _0; uint64_t _1;  };
+struct MyEnumPayloadC { uint32_t x; uint8_t y; };
+
+union MyEnumPayload {
+   uint32_t A;
+   MyEnumPayloadB B;
+   MyEnumPayloadC C;
+};
+
+struct MyEnum {
+    MyEnumTag tag;
+    MyEnumPayload payload;
+};
+```
+
+## Enums without a specified representation
+
+If no explicit `#[repr]` attribute is used, then the layout of most
+enums is not specified, with one crucial exception: option-like enums
+may in some cases use a compact layout that is identical to their
+payload.
+
+(Meta-note: The content in this section is not described by any RFC
+and is therefore "non-normative".)
+
+### Discriminant elision on Option-like enums
+
+**Definition.** An **option-like enum** is an enum which has:
+
+- one variant with a single field,
+- other variants with no fields ("unit" variants).
+
+The simplest example is `Option<T>` itself, where the `Some` variant
+has a single field (of type `T`), and the `None` variant has no
+fields.  But other enums that fit that same template (and even enums
+that include multiple `None`-like fields) fit. 
+
+**Definition.** The **payload** of an option-like enum is the single
+field which it contains; in the case of `Option<T>`, the payload has
+type `T`.
+
+**Definition.** In some cases, the payload type may contain illegal
+values, which are called **niches**. For example, a value of type `&T`
+may never be NULL, and hence defines a niche consisting of the
+bitstring `0`.  Similarly, the standard library types [`NonZeroU8`]
+and friends may never be zero, and hence also define the value of `0`
+as a niche. (Types that define niche values will say so as part of the
+description of their representation invariant.)
+
+[`NonZeroU8`]: https://doc.rust-lang.org/std/num/struct.NonZeroU8.html
+
+**Option-like enums where the payload defines an adequate number of
+niche values are guaranteed to be represented without using any
+discriminant at all.** This is called **discriminant elision**.  If
+discriminant elision is in effect, then the layout of the enum is
+equal to the layout of its payload.
+
+The most common example is that `Option<&u8>` can be represented as an
+nullable `&u8` reference -- the `None` variant is then represented
+using the niche value zero. This is because a valid `&u8` value can
+never be zero, so if we see a zero value, we know that this must be
+`None` variant.
+
+In order for the optimization to apply, the payload type must define a
+number of niches greater than or equal to the number of unit variants.
+In the case of `Option<T>`, this means that any niche at all will
+suffice, as there is only one unit variant (`None`).
+
```

Comment:
there is no stable way yet, but the compiler supports multiple niche values, and e.g. rustc takes advantage of this (all of our index types define values `0xFFFFFFF0` and above as niches, for example, so that they can be readily combined with enums)

Reply:
Which indexing types do you mean? `Option<usize>` takes 128 bits according to the playground.

Reply:
Types like `NodeId`, `CrateId` etc. are defined using [this macro](https://github.com/rust-lang/rust/blob/4755e2f3b635ad7391ba88b18672b05cbae5ba4c/src/librustc_data_structures/indexed_vec.rs#L67) which uses the `rustc_layout_scalar_valid_range_end` attribute to create a niche.